### PR TITLE
AP_ESC_Telem: include header that adds NUM_SERVO_CHANNELS

### PR DESCRIPTION
**Pull request by joshanne**
The define for `NUM_SERVO_CHANNELS` is not necessarily present, and periph could have for example, 2 SRV outputs, but with 16 max escs. This results in a `_telem_data` that could contain uninitialized/unwriteable values, and ends up in outputting bad data as _valid_ ESCs on the CAN bus.

The intention of the define block at line 8 is to make the number of available esc telems equal to the number of servo outputs. But `AP_ESC_Telem.h` is unable to see the define for `NUM_SERVO_CHANNELS` so it defaults to 16.

Simply, the logic is:
- If `NUM_SERVO_CHANNELS` is available to use... use it.
- If it's not, and we are a `AP_Periph`, with `HAL_PWM_COUNT` available, then use that.
- Otherwise, if we're a straight `AP_Periph`, we have no esc telem.
- Otherwise if minimize features is turned off and we have the flash size available, then define as 32.
- Otherwise default to 16.

### CODE REVIEW

**Review by andyp1per — APPROVED** [review]

**Comment by joshanne** [comment]
RE. the failure in sitl_periph_gps:

In theory, `HAL_WITH_ESC_TELEM` should never be defined unless there is `NUM_SERVO_CHANNELS` and/or `HAL_PWM_COUNT`. So there should be no need to define `ESC_TELEM_MAX_ESCS` as 0 here.

We should probably update the condition in `libraries/AP_ESC_Telem/AP_ESC_Telem_Backend.h` to look for `HAL_PWM_COUNT` rather than `NUM_SERVO_CHANNELS` since the define for `NUM_SERVO_CHANNELS` isn't included here either.

Do we straight swap it to `HAL_PWM_COUNT`? Or do we do a tiered approach there as well? 

So logic is:
- User specified a servo channel count, rely on that.
- User didn't specify a servo channel count, but specified PWMs, so enable it (may conflict with existing AP_Periph dependent definition)
- If not explicitly defining esc telem, then enable it if condition is met.

```
#if defined(NUM_SERVO_CHANNELS) && NUM_SERVO_CHANNELS == 0
#define HAL_WITH_ESC_TELEM 0
#if defined(HAL_PWM_COUNT) && HAL_PWM_COUNT == 0
#define HAL_WITH_ESC_TELEM 0
#elif !defined(HAL_WITH_ESC_TELEM)
#define HAL_WITH_ESC_TELEM (HAL_SUPPORT_RCOUT_SERIAL || HAL_MAX_CAN_PROTOCOL_DRIVERS) && !defined(HAL_BUILD_AP_PERIPH)
#endif
```

File: libraries/AP_ESC_Telem/AP_ESC_Telem.h [hunk 1]
```diff
@@ -9,6 +9,10 @@
 
 #ifdef NUM_SERVO_CHANNELS
  #define ESC_TELEM_MAX_ESCS NUM_SERVO_CHANNELS
+#elif defined(HAL_BUILD_AP_PERIPH) && defined(HAL_PWM_COUNT)
```
Comment (tridge) [inline libraries/AP_ESC_Telem/AP_ESC_Telem.h:12]: better to add a SRV_Channel/SRV_Channel_config.h

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_ESC_Telem/AP_ESC_Telem.h (hunk 1 above)
Comment (joshanne) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.h]: Added `SRV_Channel/SRV_Channel_config.h` and included where applicable.

**Comment by joshanne** [comment]
- [ ] Validate changes on AP_Periph hardware and CubeOrangePlus to ensure no strange defects

**Comment by joshanne** [comment]
Could the `sitltest-plane` be re-triggered? Looks like the test should have passed.

**Review by tridge — APPROVED** [review]

File: libraries/SRV_Channel/SRV_Channel_config.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+#ifndef NUM_SERVO_CHANNELS
+#if defined(HAL_BUILD_AP_PERIPH) && defined(HAL_PWM_COUNT)
```
Comment (magicrub) [inline libraries/SRV_Channel/SRV_Channel_config.h:6]: You're in-between two styles making it a little unintuitive, please pick one:

```
#ifndef NUM_SERVO_CHANNELS
#if defined(HAL_BUILD_AP_PERIPH) && defined(HAL_PWM_COUNT)
    #define NUM_SERVO_CHANNELS HAL_PWM_COUNT
#elif defined(HAL_BUILD_AP_PERIPH)
    #define NUM_SERVO_CHANNELS 0
#elif !HAL_MINIMIZE_FEATURES && BOARD_FLASH_SIZE > 1024
    #define NUM_SERVO_CHANNELS 32
#else
    #define NUM_SERVO_CHANNELS 16
#endif
#endif
```
 --- or ---
 ```
#ifndef NUM_SERVO_CHANNELS
#if defined(HAL_BUILD_AP_PERIPH)
    #if defined(HAL_PWM_COUNT)
        #define NUM_SERVO_CHANNELS HAL_PWM_COUNT
    #else
        #define NUM_SERVO_CHANNELS 0
    #endif
#else
    #if !HAL_MINIMIZE_FEATURES && BOARD_FLASH_SIZE > 1024
        #define NUM_SERVO_CHANNELS 32
    #else
        #define NUM_SERVO_CHANNELS 16
    #endif
#endif
 ```

Comment (peterbarker) [reply libraries/SRV_Channel/SRV_Channel_config.h:6]: @magicrub I was planning on coming back through here and moving those AP_PERIPH bits down into `chibios_hwdef.py`.

Given @joshanne is just moving this code from one place to another, are you OK with leaving the style as-is for now?  (i.e. the formatting is not his fault! :-)

Comment (magicrub) [reply libraries/SRV_Channel/SRV_Channel_config.h:6]: @peterbarker yeah, that's fine. I didn't realize it was actually a pure copy/paste on this one. For some reason I thought it was coming from the other block that said:
```
#ifdef NUM_SERVO_CHANNELS
 #define ESC_TELEM_MAX_ESCS NUM_SERVO_CHANNELS
#elif !HAL_MINIMIZE_FEATURES && BOARD_FLASH_SIZE > 1024
 #define ESC_TELEM_MAX_ESCS 32
#else
 #define ESC_TELEM_MAX_ESCS 16
#endif
#define ESC_TELEM_MAX_ESCS NUM_SERVO_CHANNELS```

**Review by magicrub — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**